### ISSUE TITLE
Feature Request: Enables cores to influence scaling and cropping

### ISSUE THREAD

**Issue by DerKoun** [libretro/RetroArch#10930]
For some 2D cores it would useful to be able to request that their output is scaled in specific ways.
Values in the "scaleHints" object could be:

- "interpoaltion"
-- "default": no preference
-- "enbaled": use bilinear (or better) interpolation
-- "disaled": use neares neighbor scaling
-- "down": like "enabled" when source is larger than target, like "disabled" otherwise

- "imageFit"
-- "default": no preference
-- "contain": current non-integer behavior (respect AR, maxium size without overflow, centered)
-- "cover": respect AR, minimumsize covering the entirescreen, centered, crop overflow (this may not have use-cases here)
-- "stretch": fill screen exactly, ignoring aspect ratio
-- "integerHeight"/"integerWidth": like "contain", but scale factor of height/witdh is reduced until it is integer and the one of the other dimension is adapted to keep the aspect ratio
-- "pixelPerfect": like "integerHeight"/"integerWidth", but both scale factors are reduced to the nearest integer, ignoring aspect ratio

When both source and target have a pixel aspect ratio of 1 "integerHeight", "integerWidth" & "pixelPerfect" produce the same result.

- "crop"/"softCrop"
-- "top", "bottom", "left", "right": Specifies an amoutnt of source pixel rows or columns to be ignored when scaling. This basicly reduces the dimension of the source from the four sides. The difference is that "crop" means the cropped edges should never be visible, even when they are on screen due to "imageFit" settings, while for "soft-crop" they should be visible in that case.

- "backgroundColor": The color used where no source is visisble, i.e. around centered sources due to "imageFit". This could be one color or multiple ones for qudrants or gradients.

This will need refining, but I tink it could be quite usefull, especially if it can be changed during runtime.

**Comment by hizzlekizzle** [comment]
Aside from cropping, where the core needs to be able to run the show, this is all frontend territory. What advantage would doing it in the core bring?

**Comment by DerKoun** [comment]
I agree that cropping (IMO specifically "softCrop") is the most important setting here, but I though I'd put in all related ideas, putting them up for discussion together, even if some may turn out to have no practical use-cases. I, of course, won't mind if only a subset is implemented, at least for now.

Before I go into the other settings, I want to make sure I explained correctly what these should do (I'm not a native speaker). I want these settings to be requests from the core to the frontend to override graphics settings, so the client can influence the final output without having to things like scaling itself.

"interpoaltion" and "imageFit" most for cores with multiple emulators or emulators with multiple graphics modes, so users can configure the core to switch the frontend settings automatically at the right point. I know this is far from 'necessary', but it would give the core a bit more (optional) control over the endresult. 

"backgroundColor" could be used like AmbiLight or similar technologies where TV shine colors on the wall around them to give the illusion of a larger image. Cores for some 2D consoles know a background color (or color array) for the frame and could optionally pass this to the front end.

"softCrop" would be very helpful when the frontEnd does integer scaling, so you can ignore irrelevant parts at the edges to get a larger scale factor, but you don't loose those parts when it's not necessary.

So, yes, there is only one part of this that I really think would benefit a lot of people, but the other ones may be interesting to look into as well, maybe some time in the future.
